test/unit/specs/store/Upload.spec.js

```diff
@@ -0,0 +1,56 @@
+import index from '../../../../src/store/modules/upload/index'
+
+const testAction = (action, args, state, expectedMutations, done) => {
+  let count = 0
+
+  // mock commit
+  const commit = (type, payload) => {
+    const mutation = expectedMutations[count]
+
+    try {
+      expect(mutation.type).to.equal(type)
+      if (payload) {
+        expect(mutation.payload).to.deep.equal(payload)
+      }
+    } catch (error) {
+      done(error)
+    }
+
+    count++
+    if (count >= expectedMutations.length) {
+      done()
+    }
+  }
+
+  // call action with mock commit and args
+  action({ commit, state }, ...args)
+
+  // check if mutation is dispatched
+  if (expectedMutations.length === 0) {
+    expect(count).to.equal(0)
+    done()
+  }
+}
+
+// create a mock file for test upload
+var f = new Blob()
+f.name = 'img1.jpg'
+f.size = 100
+f.lastModifiedDate = new Date()
+
+describe('index', () => {
+  it('uploadSingleFile', done => {
+    testAction(index.actions.uploadSingleFile, [f], {}, [
+      // mutation should been called in order from first to last
+      { type: 'SET_IS_UPLOADING', payload: true },
+      { type: 'SET_BYTES_UPLOADED', payload: 0 },
+      { type: 'SET_BYTES_REMAINING', payload: 100 },
+      { type: 'SET_BYTES_UPLOADED', payload: 100 },
+      { type: 'SET_BYTES_REMAINING', payload: 0 },
+      { type: 'SET_IS_UPLOADING', payload: false },
+      { type: 'SET_BYTES_UPLOADED', payload: 0 },
+      { type: 'SET_BYTES_REMAINING', payload: 0 },
+      { type: 'SET_CANCEL_UPLOAD', payload: false }
+    ], done)
+  })
+})
```

Comment:
Do you have any tests for upload failures?

Reply:
only way cause a failure right not is shut down wifi

Reply:
Awesome! Thanks!!

Reply:
the error handling is actually done by firebase itself, so sorry what I meant by in the last comment is that shut down wifi is the only way that I can fail the upload by my self...